### Hito3 para Determinantes del ingreso
*"Modelación descriptiva, que busca definir cuáles son los principales determinantes del objeto de estudio. En base a esta sección se podrá construir o depurar el modelo predictivo."*

####  Carga de librerias e importación de .csv obtenido en el hito anterior

In [1]:
%matplotlib inline

import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

import statsmodels.formula as sm
import statsmodels.formula.api as smf

#from graficos import *
import funciones as fun

import warnings
warnings.filterwarnings('ignore')

#plt.style.use('seaborn') # Gráficos estilo seaborn
#plt.rcParams["figure.figsize"] = (8,6) # Tamaño gráficos (5, 3)
#plt.rcParams["figure.dpi"] = 75 # resolución gráficos 100

In [2]:
df = pd.read_csv('income_ready.csv')
print(df.shape)
df.head()

(48842, 15)


age  workclass_recod  fnlwgt   educ_recod  educational-num      civstatus  \
0   25          private  226802  high-school                7  never-married   
1   38          private   89814  high-school                9        married   
2   28  state-level-gov  336951      college               12        married   
3   44          private  160323      college               10        married   
4   18              NaN  103497      college               10  never-married   

       collars relationship   race  gender  capital-gain  capital-loss  \
0  blue-collar    Own-child  Black    Male             0             0   
1  blue-collar      Husband  White    Male             0             0   
2  blue-collar      Husband  White    Male             0             0   
3  blue-collar      Husband  Black    Male          7688             0   
4          NaN    Own-child  White  Female             0             0   

   hours-per-week   region  income  
0              40  America       0  
1              50  America       0  
2              40  America       1  
3              40  America       1  
4              30  America       0

###  Modelación descriptiva

- Transformacion de variables categoricas con Binary Encoding.
- Se resetea el indice del dataframe


In [3]:
df_dummies = pd.get_dummies(df.dropna(), drop_first=True).reset_index(drop=True)
df_dummies.columns = [col.replace('-', '_') for col in df_dummies.columns]
print('Tamaño del dataframe con dummies:',df_dummies.shape)


Tamaño del dataframe con dummies: (46033, 34)


Se estima el modelo econométricos no lineal binario con Logit, usando df_dummies:

In [4]:
base = fun.mf(df_dummies, var_obj='income')             # Uso de la funcion para construir la base del modelo
model_logit = smf.logit(base, data=df_dummies).fit()    # Ajuste del modelo
model_logit.summary()

Optimization terminated successfully.
         Current function value: 0.332089
         Iterations 11


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                 income   No. Observations:                46033
Model:                          Logit   Df Residuals:                    45999
Method:                           MLE   Df Model:                           33
Date:                Sat, 09 Jul 2022   Pseudo R-squ.:                  0.4073
Time:                        11:20:57   Log-Likelihood:                -15287.
converged:                       True   LL-Null:                       -25791.
Covariance Type:            nonrobust   LLR p-value:                     0.000
===================================================================================================
                                      coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept                          -7.6740      0.309    -24.805      0.000      -8.280      -7.068
age                                 0.0254      0.001     18.931      0.000       0.023       0.028
fnlwgt                            7.33e-07   1.39e-07      5.280      0.000    4.61e-07    1.01e-06
educational_num                     0.2715      0.016     17.171      0.000       0.241       0.302
capital_gain                        0.0003   8.56e-06     37.179      0.000       0.000       0.000
capital_loss                        0.0007   3.05e-05     21.720      0.000       0.001       0.001
hours_per_week                      0.0296      0.001     22.654      0.000       0.027       0.032
workclass_recod_private            -0.4532      0.074     -6.102      0.000      -0.599      -0.308
workclass_recod_self_employed      -0.7927      0.081     -9.753      0.000      -0.952      -0.633
workclass_recod_state_level_gov    -0.5268      0.083     -6.349      0.000      -0.689      -0.364
workclass_recod_unemployed         -1.5592      0.803     -1.942      0.052      -3.133       0.014
educ_recod_elementary_school        0.1688      0.229      0.738      0.461      -0.280       0.617
educ_recod_high_school             -0.0776      0.046     -1.674      0.094      -0.168       0.013
educ_recod_preschool               -4.0234      4.039     -0.996      0.319     -11.939       3.892
educ_recod_university               0.0907      0.060      1.509      0.131      -0.027       0.209
civstatus_married                   0.7768      0.136      5.715      0.000       0.510       1.043
civstatus_never_married            -0.4767      0.071     -6.669      0.000      -0.617      -0.337
civstatus_separated                -0.0533      0.133     -0.401      0.688      -0.313       0.207
civstatus_widowed                   0.0791      0.127      0.621      0.535      -0.170       0.329
collars_others                     -0.7209      0.085     -8.452      0.000      -0.888      -0.554
collars_white_collar                0.6021      0.035     17.176      0.000       0.533       0.671
relationship_Not_in_family         -0.9811      0.131     -7.477      0.000      -1.238      -0.724
relationship_Other_relative        -1.4663      0.184     -7.957      0.000      -1.828      -1.105
relationship_Own_child             -1.9446      0.163    -11.900      0.000      -2.265      -1.624
relationship_Unmarried             -1.2129      0.148     -8.196      0.000      -1.503      -0.923
relationship_Wife                   1.0555      0.083     12.679      0.000       0.892       1.219
race_Asian_Pac_Islander             0.6153      0.214      2.872      0.004       0.195       1.035
race_Black                          0.2602      0.189      1.379      0.168      -0.110       0.630
race_Other                          0.1926      0.268      0.719      0.472      -0.333       0.718
race_White                          0.513

#### Primer analisis de resultados:

- El <code>Pseudo R-squ es de 0.4073</code>. Con lo cual el modelo actual explica casi el 41% de la VO (income).

- El <code>Log-Likelihood ratio indica que el modelo es significativo</code>, rechazando la hipótesis nula que no existen variables significativas. Esto se complementa con los resultados de máxima verosimilitud.

- Por medio de los intervalos de confianza([0.025 0.975]) las siguientes variables no son significativas, ya que pasan por cero:<code> workclass_recod_unemployed, educ_record (todas sus clases), civstatus_separated, civstatus_widowed, race_Black, race_Other, region_Asia, y region_Europe</code>.

Eliminaremos las variables no significativas del modelo:

In [5]:
clear_var_ns = ['workclass_recod_unemployed', 'educ_recod_elementary_school', 'educ_recod_high_school', 'educ_recod_preschool', 'educ_recod_university', 'civstatus_separated', 'civstatus_widowed', 'race_Black', 'race_Other', 'region_Asia', 'region_Europe']
df_mod = df_dummies.drop(clear_var_ns, axis=1)
print('Tamaño del dataframe eliminando variables no significativas:',df_mod.shape)

Tamaño del dataframe eliminando variables no significativas: (46033, 23)


Se realiza la nueva modelación con el df modificado:

In [6]:
base = fun.mf(df_mod, var_obj='income')             # Uso de la funcion para construir la base del modelo
model_logit2 = smf.logit(base, data=df_mod).fit()   # Ajuste del modelo
model_logit2.summary()

Optimization terminated successfully.
         Current function value: 0.332362
         Iterations 9


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                 income   No. Observations:                46033
Model:                          Logit   Df Residuals:                    46010
Method:                           MLE   Df Model:                           22
Date:                Sat, 09 Jul 2022   Pseudo R-squ.:                  0.4068
Time:                        11:20:57   Log-Likelihood:                -15300.
converged:                       True   LL-Null:                       -25791.
Covariance Type:            nonrobust   LLR p-value:                     0.000
===================================================================================================
                                      coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept                          -7.7315      0.209    -37.042      0.000      -8.141      -7.322
age                                 0.0256      0.001     19.346      0.000       0.023       0.028
fnlwgt                           7.485e-07   1.38e-07      5.422      0.000    4.78e-07    1.02e-06
educational_num                     0.2958      0.007     41.308      0.000       0.282       0.310
capital_gain                        0.0003   8.55e-06     37.199      0.000       0.000       0.000
capital_loss                        0.0007   3.05e-05     21.729      0.000       0.001       0.001
hours_per_week                      0.0296      0.001     22.653      0.000       0.027       0.032
workclass_recod_private            -0.4338      0.074     -5.889      0.000      -0.578      -0.289
workclass_recod_self_employed      -0.7735      0.081     -9.585      0.000      -0.932      -0.615
workclass_recod_state_level_gov    -0.5064      0.082     -6.146      0.000      -0.668      -0.345
civstatus_married                   0.7720      0.133      5.785      0.000       0.510       1.034
civstatus_never_married            -0.4707      0.068     -6.956      0.000      -0.603      -0.338
collars_others                     -0.7174      0.085     -8.417      0.000      -0.885      -0.550
collars_white_collar                0.6126      0.035     17.635      0.000       0.545       0.681
relationship_Not_in_family         -0.9835      0.131     -7.510      0.000      -1.240      -0.727
relationship_Other_relative        -1.4742      0.184     -7.998      0.000      -1.835      -1.113
relationship_Own_child             -1.9490      0.163    -11.951      0.000      -2.269      -1.629
relationship_Unmarried             -1.2166      0.148     -8.237      0.000      -1.506      -0.927
relationship_Wife                   1.0453      0.083     12.608      0.000       0.883       1.208
race_Asian_Pac_Islander             0.2308      0.099      2.328      0.020       0.036       0.425
race_White                          0.2885      0.057      5.086      0.000       0.177       0.400
gender_Male                         0.7782      0.063     12.323      0.000       0.654       0.902
region_Unknown                     -0.2800      0.107     -2.619      0.009      -0.490      -0.070
===================================================================================================
"""

#### Segundo analisis de resultados

Este modelo es mas adecuado debido a:

- El <code>Pseudo R-squ es de 0.4068</code>. Con lo cual el modelo sigue explicando casi el 41% de la VO (income).
- El Log-Likelihood ratio sigue indicando que el modelo es significativo,<code> rechazando la hipótesis nula </code>(que no existen variables significativas).

#### Definición de las estrategias de Modelación predictiva:

Luego de la revision de la data y de la modelación descriptiva, se realizará una predicción mediante regresión logística.

- Se usará un módulo de <code>regresión logística</code> de la librería <code>scikit-learn</code>.
    - Se generarán conjuntos de entrenamiento y validación.
    - Se estandarizarán los datos ya que existen variables con valores que podrían sesgar el modelo (por ejemplo "fnlwgt").
    - Se ajustarán los datos, se realizarán predicciones y se compararán con nuevos modelos generados con los resultados que se vayan generando.
    

Se usará el dataframe obtenido del segundo analisis de este hito (<code>df_mod </code>, con las variables depuradas obtenidas con Logit).

In [7]:
#Generacion de .csv procesado
df_mod.to_csv('income_mod.csv', index=False)